### PR TITLE
fix(archiver): archiver migration to archives

### LINKED ISSUE #3051: Explore future of mholt/archiver
### Describe what should be investigated or refactored

We are currently using a mix of stdliib tar and https://github.com/mholt/archiver to manage tar archives. The archiver lib has a lot of good features to partially unpack tar files. The latest release of the lib was made in November 2021 with work for a new major release still in progress.

There are a couple of small bugs that we have found in mholt/archiver when run on Windows. Considering the lack of releases we should consider the future of using the lib, and consider if we should only use the stdlib or some other lib.

### PR DESCRIPTION
**Pull request by brandtkeller**
## Description

Migration away from `mholt/archiver` to mholt/archives`. 

Looked at using the go stdlib and believe that it doesn't support the same archive extensions for compression options. Particularly I did not find `.zst` support which is common in zarf package workflows. 

## Related Issue

Fixes #3051

Relates to #

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [ ] [Contributor Guide Steps](https://github.com/zarf-dev/zarf/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

File: src/pkg/packager/sources/utils.go [hunk 1]
```diff
@@ -36,46 +37,72 @@ func IsValidFileExtension(filename string) bool {
...
 	}
-	_, ok = format.(*archiver.Tar)
-	if ok {
-		return tb, nil
+	// else if ok {
+	// 	return path, nil
+	// }
+
+	// 3) try each extension in order
+	for _, ext := range []string{".tar.zst", ".tar.gz", ".tar.xz", ".tar"} {
```
Comment (mkcp) [inline src/pkg/packager/sources/utils.go:86]: Similar magic-string comment as in `archive` but in this case it makes sense to me to keep this string range local to the function `identifyUnknownTarball`.

File: src/pkg/archive/archive_test.go [hunk 2]
```diff
@@ -5,48 +5,197 @@
...
+	archivePath2 := filepath.Join(tmp, "archive2.tar.gz")
+	require.NoError(t, Compress(ctx, []string{origFile}, archivePath2, CompressOpts{}))
+
+	// Decompress with OverwriteExisting
+	opts := DecompressOpts{OverwriteExisting: true}
+	require.NoError(t, Decompress(ctx, archivePath2, outDir, opts))
+	if got := readFile(t, outFile); got != "new content" {
+		t.Errorf("expected overwritten file to have 'new content', got %q", got)
+	}
+}
```
Comment (mkcp) [inline src/pkg/archive/archive_test.go:201]: In an ideal world these tests would be expressed using the table driven tests pattern: https://go.dev/wiki/TableDrivenTests#parallel-testing. This PR does fix a CVE though, so I don't see this as a blocking request, just a quality of life improvement for maintainers down the line if we can budget the time. Thoughts?

**Review by mkcp — CHANGES_REQUESTED** [review]
Overall direction is good, some nitpicks and one discussion item about table driven tests on `archive`. Table driven tests aren't a blocking concern but something we should deliver if we can.

File: src/pkg/archive/archive.go [hunk 3]
```diff
@@ -13,48 +13,334 @@ import (
...
 
 // DecompressOpts provides optional parameters for Decompress
 type DecompressOpts struct {
-	// UnarchiveAll, when enabled, walks the sourceArchive and unarchives everything at the root of the archive.
-	// NOTE(mkcp): This is equivalent to a recursive walk with depth 1.
+	// UnarchiveAll walks root of the archive and unpacks nested .tar files.
 	UnarchiveAll bool
+
+	// Files, if non-empty, means "only extract these exact archive-paths."
+	Files []string
```
[the remark this answers is not in the document]
Comment (brandtkeller) [reply src/pkg/archive/archive.go:127]: It is a really tricky subject (I also don't have a windows test box) but we do run our unit and non-cluster E2E tests against windows and none of them are failing currently... Are there changes you would like to see?

[the remark this answers is not in the document]
Comment (mkcp) [reply src/pkg/archive/archive.go:127]: No changes requested here, just wanted to raise it as a consideration!

File: src/pkg/archive/archive.go [hunk 4]
```diff
@@ -13,59 +13,360 @@ import (
...
 	"strings"
 
-	"github.com/mholt/archiver/v3"
 	"github.com/mholt/archives"
 	"github.com/zarf-dev/zarf/src/config/lang"
-	"github.com/zarf-dev/zarf/src/pkg/layout"
 )
 
-const rwxPerm = 0o755
+const (
```
Comment (mkcp) [inline src/pkg/archive/archive.go:20]: clean 🫰

File: src/pkg/archive/archive_test.go [hunk 5]
```diff
@@ -5,48 +5,243 @@
...
+
+func TestCompressAndDecompress_MultipleFormats(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+
+	formats := []struct {
+		name      string
+		extension string
+	}{
+		{"tar", extensionTar},
```
Comment (mkcp) [inline src/pkg/archive/archive_test.go:48]: 🔥

**Review by mkcp — DISMISSED** [review]
LGTM

File: src/pkg/archive/archive.go [hunk 6]
```diff
@@ -13,48 +13,334 @@ import (
...
+	for _, src := range sources {
+		mapping[src] = filepath.Base(src)
+	}
+	files, err := archives.FilesFromDisk(ctx, nil, mapping)
+	if err != nil {
+		return fmt.Errorf("failed to stat sources: %w", err)
+	}
+
+	// Pick formatter based on extension
+	switch {
```
[the remark this answers is not in the document]
Comment (mkcp) [reply src/pkg/archive/archive.go]: 💯 always in favor of expressing code in terms of dataflow (declarative) rather than control flow (imperative). e.g. https://slikts.github.io/concurrency-glossary/?id=control-flow-vs-dataflow-imperative-vs-declarative

**Review by mkcp — APPROVED** [review]

**Review by AustinAbro321 — CHANGES_REQUESTED** [review]
A few things to clean up, otherwise looks good. Nice work

File: src/internal/packager2/layout/package.go [hunk 7]
```diff
@@ -140,18 +149,62 @@ func (e *NoSBOMAvailableError) Error() string {
...
 	}
-	path := filepath.Join(destPath, p.Pkg.Metadata.Name)
-	// TODO(mkcp): See https://github.com/zarf-dev/zarf/issues/3051
-	err := archiver.Extract(filepath.Join(p.dirPath, SBOMTar), "", path)
+
+	// 1) locate the sboms archive under the layout directory
+	sbomArchive := filepath.Join(p.dirPath, SBOMTar)
+
+	// 2) mount it as a virtual filesystem
+	ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (brandtkeller) [reply src/internal/packager2/layout/package.go]: added context to `GetSBOM` and the `GetComponentDir` methods.

File: src/internal/packager2/layout/package.go [hunk 8]
```diff
@@ -169,9 +222,7 @@ func (p *PackageLayout) GetComponentDir(destPath, componentName string, ct Compo
 		return "", err
 	}
 	defer os.RemoveAll(tmpDir)
-	// TODO (phillebaba): We are not using archiver.Extract here because there is a bug in Windows where the files will not be extracted properly from nested directories.
-	// https://github.com/zarf-dev/zarf/issues/3051
-	err = archiver.Unarchive(sourcePath, tmpDir)
+	err = archive.Decompress(context.Background(), sourcePath, tmpDir, archive.DecompressOpts{})
```
[the remark this answers is not in the document]
Comment (brandtkeller) [reply src/internal/packager2/layout/package.go]: Added

File: src/pkg/archive/archive.go [hunk 9]
```diff
@@ -13,59 +13,332 @@ import (
...
 			return err
 		}
-		if strings.HasSuffix(path, ".tar") {
-			dst := filepath.Join(strings.TrimSuffix(path, ".tar"), "..")
+		if strings.HasSuffix(path, extensionTar) {
+			dst := filepath.Join(strings.TrimSuffix(path, extensionTar), "..")
 			// Unpack sboms.tar differently since it has a different folder structure than components
-			if info.Name() == layout.SBOMTar {
-				dst = strings.TrimSuffix(path, ".tar")
+			if info.Name() == sbomFileName {
```
[the remark this answers is not in the document]
Comment (brandtkeller) [reply src/pkg/archive/archive.go:340]: I agree with the removal - it wasn't a key piece I looked at closely but will tag for a follow on PR

File: src/pkg/archive/archive.go [hunk 10]
```diff
@@ -13,59 +13,332 @@ import (
...
+// into dst, and errors if any one of them was not found.
+func unarchiveFiltered(ctx context.Context, src, dst string, want []string, skipValidation bool) error {
+	file, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("unable to open archive %q: %w", src, err)
+	}
+	defer func() {
+		cErr := file.Close()
+		err = errors.Join(err, cErr)
+	}()
```
[the remark this answers is not in the document]
Comment (brandtkeller) [reply src/pkg/archive/archive.go:251]: added

File: src/pkg/archive/archive_test.go [hunk 11]
```diff
@@ -5,48 +5,243 @@
...
+		name        string
+		setup       func(t *testing.T, ctx context.Context) (archivePath, outDir string, opts DecompressOpts)
+		expectError string
+		verify      func(t *testing.T, outDir string)
 	}{
 		{
-			name: "CompressOpts can be empty",
-			opts: CompressOpts{},
+			name: "Filtered_NotFound",
+			setup: func(t *testing.T, ctx context.Context) (string, string, DecompressOpts) {
```
[the remark this answers is not in the document]
Comment (brandtkeller) [reply src/pkg/archive/archive_test.go:115]: Agreed

File: src/pkg/archive/archive.go [hunk 12]
```diff
@@ -13,59 +13,332 @@ import (
...
+	extensionTbr  = ".tbr"
+	extensionLz4  = ".tar.lz4"
+	extensionTlz4 = ".tlz4"
+	extensionLzip = ".tar.lz"
+	extensionMz   = ".tar.mz"
+	extensionTmz  = ".tmz"
+	sbomFileName  = "sbom.tar"
+
+	dirPerm  = 0o755
+	filePerm = 0o644
```
[the remark this answers is not in the document]
Comment (AustinAbro321) [reply src/pkg/archive/archive.go:42]: That's a fair argument and something I hadn't thought of. I can see the benefits for both, and don't feel strongly either way.

File: src/internal/packager2/layout/package.go [hunk 13]
```diff
@@ -140,18 +149,62 @@ func (e *NoSBOMAvailableError) Error() string {
...
-	path := filepath.Join(destPath, p.Pkg.Metadata.Name)
-	// TODO(mkcp): See https://github.com/zarf-dev/zarf/issues/3051
-	err := archiver.Extract(filepath.Join(p.dirPath, SBOMTar), "", path)
+
+	// 1) locate the sboms archive under the layout directory
+	sbomArchive := filepath.Join(p.dirPath, SBOMTar)
+
+	// 2) mount it as a virtual filesystem
+	ctx := context.Background()
+	fsys, err := archives.FileSystem(ctx, sbomArchive, nil)
```
[the remark this answers is not in the document]
Comment (brandtkeller) [reply src/internal/packager2/layout/package.go]: I believe we can now that some of the original work was restructured
